### PR TITLE
ASF migration apigateway

### PR DESCRIPTION
**Pull request by calvernaz**
## Verification steps

```
 awslocal apigateway create-rest-api --name "demo-api"
```

then

```
awslocal apigateway get-rest-api --rest-api-id=<rest-api-id>
```

### CODE REVIEW

**Comment by whummer** [comment]
Cool, thanks for pushing this topic forward @calvernaz @thrau !

I've rebased your branch onto latest `master`, and played around a bit with an alternative approach that uses a generic fallback listener `CallMotoListener`. We'd create the `Service` with an `AsfWithFallbackListener` that calls an instance of `CallMotoListener` whenever a method is not directly implemented in the provider (i.e., raises `NotImplementedError`).

Just a PoC for now, and lots of corner cases to consider, but would be great to get your thoughts on this.

**Comment by thrau** [comment]
that looks very useful! i can see that there are multiple streams of work now trying to solve the same problem, namely reconciling ASF and existing service listeners, which sometimes proxy to flask, moto, or are implemented from scratch. it absolutely makes sense to have multiple solutions available for different cases.

i think it makes sense to start consolidating the works from also @dfangl in the lambda branch where he had a similar problem.

File: localstack/aws/proxy.py [hunk 1]
```diff
@@ -70,6 +74,55 @@ def _raise_not_implemented_error(*args, **kwargs):
...
+
+    def _convert_moto_response(self, response):
+        # taken from moto.core.responses
+        if response is None:
+            response = "", {}
+        if len(response) == 2:
+            body, new_headers = response
+        else:
+            status, new_headers, body = response
+        status = new_headers.get("status", 200)
```
Comment (whummer) [inline localstack/aws/proxy.py:122]: This logic is incorrect, still needs to be fixed...

**Comment by calvernaz** [comment]
> Cool, thanks for pushing this topic forward @calvernaz @thrau !
> 
> I've rebased your branch onto latest `master`, and played around a bit with an alternative approach that uses a generic fallback listener `CallMotoListener`.
> 
> The idea is basically: We'd create the `Service` with an `AsfWithFallbackListener` that calls an instance of `CallMotoListener` whenever a method is not directly implemented in the provider (i.e., raises `NotImplementedError`).
> 
> One use case that is not yet covered here is to call the moto backend AND then patching its response before returning it to the client. For these cases, we could implement the logic in the ASF provider's method - make an explicit call to the moto backend (with the appropriate region) to get the response, and then patching/modifying the response before returning it.
> 
> Just a PoC for now, and lots of corner cases to consider, but would be great to get your thoughts on this.



> Cool, thanks for pushing this topic forward @calvernaz @thrau !
> 
> I've rebased your branch onto latest `master`, and played around a bit with an alternative approach that uses a generic fallback listener `CallMotoListener`.
> 
> The idea is basically: We'd create the `Service` with an `AsfWithFallbackListener` that calls an instance of `CallMotoListener` whenever a method is not directly implemented in the provider (i.e., raises `NotImplementedError`).
> 
> One use case that is not yet covered here is to call the moto backend AND then patching its response before returning it to the client. For these cases, we could implement the logic in the ASF provider's method - make an explicit call to the moto backend (with the appropriate region) to get the response, and then patching/modifying the response before returning it.
> 
> Just a PoC for now, and lots of corner cases to consider, but would be great to get your thoughts on this.

I think the biggest advantage of relying on the moto mocking and botocore is the boundary is defined by the client, boto, which is the official implementation and not importing any internals into localstack.

With this alternative, we are importing the internals of moto handlers into localstack, but I guess there is no real danger because the version is pinned and won't suddenly break.

**Comment by whummer** [comment]
Yep, good points @calvernaz . Also briefly discussed with @thrau just now - agreed that having the boundary via botocore will be beneficial. 👍 If we manage to create a mock `boto3` client that calls `moto` under the covers, **without** using the `BotocoreStubber` in moto (as this one uses global state in the `BUILTIN_HANDLERS` botocore handlers), that would be a great path forward! 👍

**Comment by calvernaz** [comment]
> Yep, good points @calvernaz . Also briefly discussed with @thrau just now - agreed that having the boundary via botocore will be beneficial. 👍 If we manage to create a mock `boto3` client that calls `moto` under the covers, **without** using the `BotocoreStubber` in moto (as this one uses global state in the `BUILTIN_HANDLERS` botocore handlers), that would be a great path forward! 👍
> 
> (PS: Sorry for polluting your branch with the changes 😄 , but I think this is a nice way to collaborate and iterate on the different ideas..)

Not polluting it all, chime in, all of you! This is from what I stand a core how these interactions work. You mentioned the issue when we want to use boto to call back w/o going to moto. Is there a way to not do that? Any examples where we do that?

**Comment by thrau** [comment]
maybe this can be a way forward to transparently call boto from ASF without having to do the call explicitly in the ASF provider.

some clarifications:
* this translate a call to ASF to a boto client call, so as to proxy ASF calls to arbitrary backends that are abstracted by boto
* so the call path is: `:4566` -> ASF -> `CallBotoHandler` -> boto (great abstraction and API boundary) -> (HTTP) -> moto, AWS, other localstack providers, wherever....
  * at some point, a modified boto client can skip the HTTP call and call the target directly (this may prove to be non-trivial depending on the target)
  * `botocore.client.BaseClient._make_request` or `_make_api_call` i think are the correct functions to modify.
  * these modified boto clients can also be used programmatically in ASF provider implementations to make calls explicitly rather than magically through the `CallBotoHandler`.
* what `main` does in the example below is basically what happens inside the `Skeleton` to call the correct ASF provider function

```python
from typing import Optional

from localstack.aws.api import RequestContext, ServiceRequest, ServiceResponseto skip the http call 
from localstack.aws.api.sqs import CreateQueueRequest
from localstack.aws.skeleton import CallBotoHandler
from localstack.aws.spec import load_service
from localstack.utils.aws import aws_stack


class CallBotoHandler:
    """A ServiceRequestHandler that forward to call to the appropriate boto client."""

    def __init__(self, endpoint_url: str = None) -> None:
        super().__init__()
        self.endpoint_url = endpoint_url

    def __call__(
            self, context: RequestContext, request: ServiceRequest
    ) -> Optional[ServiceResponse]:
        client = self._get_boto_client(context)
        # this is the generic variant of, e.g., sqs_client.create_queue(QueueName="myqueue")
        result = client._make_api_call(context.operation.name, request)
        return result

    def _get_boto_client(self, context: RequestContext):
        return aws_stack.connect_to_service(
            context.service.service_name, endpoint_url=self.endpoint_url, region_name=context.region
        )


def main():
    # build request context
    context = RequestContext()
    context.region = "us-east-1"
    context.service = load_service("sqs")
    context.operation = context.service.operation_model("CreateQueue")

    # build request instance
    request: ServiceRequest = CreateQueueRequest(
        QueueName="myqueue"
    )

    # this is an ASF request handler that calls boto underneath
    handler = CallBotoHandler()
    handler(context, request)


if __name__ == '__main__':
    main()

```

**Pull request closed without merging**